File: jafu/src/main/java/org/springframework/fu/jafu/ConfigurationDsl.java
package org.springframework.fu.jafu;

import java.util.function.Consumer;

import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.boot.context.properties.FunctionalConfigurationPropertiesBinder;
import org.springframework.boot.context.properties.bind.Bindable;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ApplicationEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.context.support.GenericApplicationContext;


/**
 * Jafu DSL for modular configuration that can be imported in the application.
 *
 * Usage: {@code Consumer<ConfigurationDsl> conf = c -> c.beans(b -> b.bean(Foo.class));}
 *
 * @see ApplicationDsl#enable(ApplicationContextInitializer)
 * @see ApplicationDsl#enable(Consumer)
 *
 * @author Sebastien Deleuze
 */
public class ConfigurationDsl extends AbstractDsl {

	private final Consumer<ConfigurationDsl> dsl;

	ConfigurationDsl(Consumer<ConfigurationDsl> dsl) {
		super();
		this.dsl = dsl;
	}

	/**
	 * Configure global, package or class logging via a {@link LoggingDsl dedicated DSL}.
	 */
	public ConfigurationDsl logging(Consumer<LoggingDsl> dsl) {
		new LoggingDsl(dsl);
		return this;
	}

	/**
	 * Specify the class and the prefix of configuration properties, which is the same mechanism than regular Boot
	 * configuration properties without `@ConfigurationProperties` annotation.
	 * @see #configurationProperties(Class, String)
	 * @see <a href="https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties">Type-safe Configuration Properties</a>
	 */
	public <T> ConfigurationDsl configurationProperties(Class<T> clazz) {
		configurationProperties(clazz, "");
		return this;
	}

	/**
	 * Specify the class and the optional prefix of configuration properties, which is the same mechanism than regular
	 * Boot configuration properties without `@ConfigurationProperties` annotation.
	 * @see #configurationProperties(Class)
	 * @see <a href="https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties">Type-safe Configuration Properties</a>
	 */
	public <T> ConfigurationDsl configurationProperties(Class<T> clazz, String prefix) {
		context.registerBean(clazz.getSimpleName() + "ConfigurationProperties", clazz, () -> new FunctionalConfigurationPropertiesBinder(context).bind(prefix, Bindable.of(clazz)).get());
		return this;
	}

	/**
	 * Configure beans via a {@link BeanDefinitionDsl dedicated DSL}.
	 */
	public ConfigurationDsl beans(Consumer<BeanDefinitionDsl> dsl) {
		new BeanDefinitionDsl(dsl).initialize(context);
		return this;
	}

	/**
	 * Enable the specified functional configuration.
	 * @see #enable(Consumer)
	 */
	@Override
	public ConfigurationDsl enable(ApplicationContextInitializer<GenericApplicationContext> configuration) {
		return (ConfigurationDsl) super.enable(configuration);
	}

	/**
	 * Enable the specified functional configuration.
	 * @see #enable(ApplicationContextInitializer)
	 */
	public ConfigurationDsl enable(Consumer<ConfigurationDsl> configuration) {
		new ConfigurationDsl(configuration).initialize(context);
		return this;
	}

	/**
	 * Declare application event Listeners in order to run tasks when {@link ApplicationEvent}
	 * like {@link ApplicationReadyEvent} are emitted.
	 */
	@SuppressWarnings("unchecked")
	public <E extends ApplicationEvent> ConfigurationDsl listener(Class<E> clazz, ApplicationListener<E> listener) {
		context.addApplicationListener(e -> {
			// TODO Leverage SPR-16872 when it will be fixed
			if (clazz.isAssignableFrom(e.getClass())) {
				listener.onApplicationEvent((E)e);
			}
		});
		return this;
	}

	@Override
	public void initialize(GenericApplicationContext context) {
		super.initialize(context);
		this.dsl.accept(this);
	}

}

File: kofu/src/main/kotlin/org/springframework/fu/kofu/ConfigurationDsl.kt
package org.springframework.fu.kofu

import org.springframework.boot.context.properties.FunctionalConfigurationPropertiesBinder
import org.springframework.boot.context.properties.bind.Bindable
import org.springframework.context.ApplicationEvent
import org.springframework.context.support.BeanDefinitionDsl
import org.springframework.context.support.GenericApplicationContext
import org.springframework.context.support.registerBean

/**
 * Kofu DSL for a configuration that can be imported in an application or used in tests.
 *
 * @see configuration
 * @see enable
 * @sample org.springframework.fu.kofu.samples.applicationDslWithConfiguration
 * @author Sebastien Deleuze
 */
open class ConfigurationDsl(private val dsl: ConfigurationDsl.() -> Unit): AbstractDsl() {

	/**
	 * Configure beans via a [dedicated DSL][BeanDefinitionDsl].
	 *
	 * Since classes with a single constructor have their parameters automatically autowired,
	 * it is recommended to use constructor injection with `val` read-only
	 * (and non-nullable when possible) private [properties](https://kotlinlang.org/docs/reference/properties.html).
	 *
	 * @sample org.springframework.fu.kofu.samples.beansDsl
	 */
	fun beans(dsl: BeanDefinitionDsl.() -> Unit) {
		BeanDefinitionDsl(dsl).initialize(context)
	}

	/**
	 * Enable the specified functional configuration.
	 * @see configuration
	 * @sample org.springframework.fu.kofu.samples.applicationDslWithConfiguration
	 */
	fun enable(configuration: AbstractDsl) {
        configuration.initialize(context)
	}

	/**
	 * Bind and return the specified configuration properties class. If you need a bean,
	 * register it via the [beans] DSL.
	 *
	 * @sample org.springframework.fu.kofu.samples.configurationProperties
	 */
	inline fun <reified T : Any> configurationProperties(properties: T? = null, prefix: String = "", defaultProperties: T? = null): T {
		val bindedProperties = properties ?: FunctionalConfigurationPropertiesBinder(context).bind(prefix, Bindable.of(T::class.java)).orElse(defaultProperties)
		context.registerBean<T>("${T::class.java.simpleName.toLowerCase()}ConfigurationProperties") {
			bindedProperties
		}
		return bindedProperties
	}

	/**
	 * Configure global, package or class logging via a [dedicated DSL][LoggingDsl].
	 * @sample org.springframework.fu.kofu.samples.loggingDsl
	 */
	fun logging(dsl: LoggingDsl.() -> Unit) {
		LoggingDsl(dsl).initialize(context)
	}

	/**
	 * Declare application event Listeners in order to run tasks when `ApplicationEvent`
	 * like `ApplicationReadyEvent` are emitted.
	 * @sample org.springframework.fu.kofu.samples.listener
	 */
	inline fun <reified E : ApplicationEvent>listener(crossinline listener: BeanDefinitionContext.(E) -> Unit) {
		context.addApplicationListener {
			// TODO Leverage SPR-16872 when it will be fixed
			if (it is E) {
				listener.invoke(BeanDefinitionContext(context), it)
			}
		}
	}

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		dsl()
	}

}
